migrating-to-cd4pe.md

```diff
@@ -0,0 +1,126 @@
+
+This change should cause no changes due to the git operations we have been following, however this can be verified by running:
+
+```shell
+git diff cd4pe_production production
+```
+
+#### Environment Groups
+
+If the user is already using environment groups, then the environment can simply be switched from the existing environment, to the new `cd4pe_` equivalent i.e. `development` would become `cd4pe_development`.
+
+This change should cause no changes due to the git operations we have been following, however this can be verified by running:
+
+```shell
+git diff cd4pe_production production
+```
+
+### Finishing Up
+
+Once all nodes are running against `cd4pe_` branches we can create a CD4PE pipeline and we are done! Since the merge-back the `master` branch is a linear history of commits, further changes should be merged into the `master` branch from feature branches. The CD4PE pipeline will then move the head of the bookkeeping branches forward along the master branch, as it is designed to. The old branches (`development`, `staging` and `production`) can now be deleted.
+
+## Things to Look Out For
+
+### Usage of `%{environment}` in hiera
+
+Since moving to CD4PE recommends that nodes change their environment, this can cause issues if the hiera hierarchy includes that node's environment. There are two ways to get around this:
+
+* Replace all instances of the old environment names with the new environment names, but only in the new branches
+* Simply do not rename the branches from `environment` to `cd4pe_environment`. CD4PE doesn't require that this change be made, but it does allow for differentiation between the book-keeping branches and those that are designed to be used by humans
+
+### Half-Merged Changes
+
+It is recommended that impact analysis is always used in a pipeline before triggering a deployment. If there were changes in flux at the time of cutover that had not yet reached the `production` branch (but had reach other branches such as `staging`) they will be deployed if fresh commits to the `master` branch are pushed through the pipeline since not all environments are at the same revision. CD4PE does not have the ability for changes to "overtake" each other in the pipeline as was possible with the previous workflow. It is best to think of the pipeline as fully loaded with changes and attempting to push a single change all the way though will also flush out any other changes that had been sitting in there. Impact analysis when used correctly will detect these and allow for informed decisions at the time of deployment.
```

Comment:
```suggestion
It is recommended that impact analysis is always used in a pipeline before triggering a deployment. If there were changes in flux at the time of cutover that had not yet reached the `production` branch (but had reached other branches such as `staging`) they will be deployed if fresh commits to the `master` branch are pushed through the pipeline since not all environments are at the same revision. CD4PE does not have the ability for changes to "overtake" each other in the pipeline as was possible with the previous workflow. It is best to think of the pipeline as fully loaded with changes and attempting to push a single change all the way though will also flush out any other changes that had been sitting in there. Impact analysis when used correctly will detect these and allow for informed decisions at the time of deployment.
```